Flag fixes for ROCm libraries

AngryLoki opened the pull request:
This adds a new function `rocm_use_hipcc` in `rocm.eclass` to be used in every `src_configure`, where compiler is switched to hipcc (a. k. a. clang):
```
	export CC=hipcc CXX=hipcc
	strip-unsupported-flags
```

Why not `if ! tc-is-clang ; then strip-unsupported-flags; fi` - because user may want to set systemwide `clang-19` with `CXXFLAGS="-fawesome-clang-19-only-flag"` in `/etc/portage/make.conf`, and ROCm needs clang-18 due to LLVM-bitcode compatibility.

Why not `if [[ -n "${have_switched_compiler}" ]] ; then strip-unsupported-flags && filter-lto; fi` because while `strip-unsupported-flags` does pretty good job, `filter-lto` is too extreme: for example it removes both `-flto` (working flag for clang) and `-flto=16` (gcc-only flag).

Also roctracer and hipBLASLt were not compatible with libc++ (in `sys-devel/clang-common[default-libcxx]` systems), added `append-cxxflags -stdlib=libstdc++` there.

Closes: https://bugs.gentoo.org/936099
Closes: https://bugs.gentoo.org/935998
Closes: https://bugs.gentoo.org/862837
Closes: https://bugs.gentoo.org/854111
Closes: https://bugs.gentoo.org/937276
Closes: https://bugs.gentoo.org/928906

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

AngryLoki commented:
@candrews , could you check this, please?

AngryLoki commented:
rebased and applied to new `dev-libs/rocm-device-libs-6.1.2`

negril commented:
That looks like a whole lot of code duplication and thus potential for errors. Maybe prepare splitting that out into `rocm.eclass` by creating two functions `rocm_use_clang` and `rocm_use_hipcc`?

> ROCm needs clang-18 due to LLVM-bitcode compatibility.

That sounds like a use-case for `LLVM_SLOT` from `llvm-r1.eclass`?

AngryLoki commented:
@negril , thank you for suggestions! I added `rocm_use_hipcc`. Decided not to add `rocm_use_clang`, as it belongs only to rocm-device-libs and should not be reused.

Regarding `LLVM_SLOT`, it is already used in ebuild, I only mentioned clang-18 in the pull-request description.

AngryLoki commented:
@littlewu2508 , as now it changes `rocm.eclass`, maybe you could find some time to check this - https://github.com/gentoo/gentoo/pull/37639/files#diff-95996e1d44358e67d1a432f1ce07178736cfd51593bea0a28c5da7fdc9153c11

littlewu2508 commented:
> @littlewu2508 , as now it changes `rocm.eclass`, maybe you could find some time to check this - https://github.com/gentoo/gentoo/pull/37639/files#diff-95996e1d44358e67d1a432f1ce07178736cfd51593bea0a28c5da7fdc9153c11

Thank you for reminding, and sorry for my absence recently. The idea of using `strip-unsupported-flags` looks good. I'll take a closer look and have some tests in the following days

littlewu2508 commented on dev-util/roctracer/roctracer-6.1.1.ebuild:
```diff
@@ -67,6 +67,11 @@ src_prepare() {
 }
 
 src_configure() {
+	rocm_use_hipcc
```
If I remember correctly roctracer can be built with gcc, since it is a host library, not containing GPU kernel.

Only when building with `test`, it compiles some hip program (to test the tracing capabilities). So my humble suggestions are (in order of ascending gracefulness):

1. `rocm_use_hipcc` only when using `test`, or
2. just use `strip-unsupported-flags` when using `test`, or
3. separate the build process of test suite, like `dev-util/hip`, but that require some patch to `test/CMakeLists.txt `

AngryLoki replied:
`rocm_use_hipcc` only when using `test` sounds not great: this technically means that emerge without tests installs completely different shared library (different size, different methods of hardened flags, maybe libstdc++ instead of libc++, potential unidentified problems if library used clang-specific code and developers never tested gcc builds due to hipcc requirement). I suggest to keep it simple/stupid; anyways, hipcc will very likely be installed by other packages.

littlewu2508 replied:
>  I suggest to keep it simple/stupid; anyways, hipcc will very likely be installed by other packages.

OK, that seems reasonable. I'll run tests to see if switching to hipcc/clang cause any issues

littlewu2508 commented:
> as now it changes `rocm.eclass`

Also the changes to rocm.eclass needs to be sent to gentoo-dev mailing list

AngryLoki commented:
> Also the changes to rocm.eclass needs to be sent to gentoo-dev mailing list

Done. Also added:
* similar fix to `dev-util/hipify-clang` - https://bugs.gentoo.org/928906
* updated commit message for `dev-libs/rocm-device-libs` to close https://bugs.gentoo.org/937276

littlewu2508 commented on eclass/rocm.eclass:
```diff
@@ -15,9 +15,13 @@
 # read and write permissions to GPU device in src_test phase, throwing friendly
-# error message if unavailable.
+# error message if unavailable. However src_configure in general should not
+# access any GPU devices. If it does, it usually means that CMakeLists.txt
```
BTW, have you observed any package that probes GPU device in `src_configure`?

negril replied:
Anything CUDA related if you don't specify an architecture.

littlewu2508 replied:
OK, that sounds reasonable. And what should we do for those packages? Add `addpredict` in the `src_configure` of their ebuilds?

negril replied:
I'm very much against addpredict. You end up with silent failures that are very hard to detect.

Either you fail hard and completely avoid the detection. Or we actually allow the detection of the native device.

AngryLoki replied:
I rechecked all issues that caused sandbox violations in the past for ROCm libraries.

The main root cause was hipcc: it automatically discovers gpu just when user calls `hipcc test.cpp` - see https://wiki.gentoo.org/wiki/HIP#hipcc_.28Clang_wrapper.29 . This call happens every time when cmake tests that C/C++ "is able to compile a simple test program". However, many ebuilds use hipcc without addpredict now and almost nobody sees sandbox violation errors is because of improvements in amd_rocm_enumerator implementation. 

As described in https://github.com/ROCm/rocminfo?tab=readme-ov-file#execution, enumeration consists of 4 steps:
1) read file from ROCM_TARGET_LST env variable (if any)
2) call `cat /sys/class/kfd/kfd/topology/nodes/*/properties | grep gfx`
3) call `/usr/bin/lspci -n -d 1002:` and try to match with a dictionary
4) finally, if nothing was found during 1-3, call rocminfo (which causes sandbox violation, if kfd device exists)

After all recent changes, autodetection still happens, but it does not cause sandbox errors because execution never reaches step 4. However we can do better and effectively disable 2 and 3 with `ROCM_TARGET_LST` file. I updated `rocm_use_hipcc` adding the following:

```bash
	local target_lst="${BUILD_DIR}"/gentoo_rocm_target.lst
	if [[ "${AMDGPU_TARGETS[@]}" = "" ]]; then
		# Expected no GPU code; still need to calm down sandbox
		echo "gfx000" > "${target_lst}"
	else
		printf "%s\n" ${AMDGPU_TARGETS[@]} > "${target_lst}"
	fi
	export ROCM_TARGET_LST="${target_lst}"
```

With this code I was able to safely remove addpredict in src_configure from rocBLAS and rocSPARSE (and in 6.1.x release there are no more ROCm packages with addpredict).

Note that some other ROCm libraries (not in Gentoo) still call rocminfo (which requires addpredict) without specific reason:
* https://github.com/ROCm/rocAL/blob/rocm-6.2.0/rocAL-setup.py#L80
* https://github.com/ROCm/rocDecode/blob/rocm-6.2.0/rocDecode-setup.py#L63

But for now I don't see that anybody uses these libraries, so no ebuilds for them.

littlewu2508 commented on eclass/rocm.eclass:
```diff
@@ -231,3 +235,27 @@ check_amdgpu() {
+	# Since CMake checker does not specify --offload-arch=, hipcc enumerates devices using four methods
+	# until it finds at least one device. Last way is by accessing them (via rocminfo).
+	# To prevent potential sandbox violations, we set the ROCM_TARGET_LST variable (which is checked first).
+	local target_lst="${BUILD_DIR}"/gentoo_rocm_target.lst
```
Maybe use temp dir ${T} would be better?

```suggestion
	local target_lst="${T}"/gentoo_rocm_target.lst
```

AngryLoki replied:
Hm, I think BUILD_DIR is better. `gentoo_rocm_target.lst` now resides near `gentoo_common_config.cmake`, `gentoo_rules.cmake` and `gentoo_toolchain.cmake`.

littlewu2508 replied:
Well, changing to `${T}` can avoid the problem of nonexistent `BUILD_DIR`. In the future there may be other build system other than cmake calling hipcc, and their ebuild may not use cmake.eclass and have `BUILD_DIR` set.

littlewu2508 commented on dev-util/Tensile/Tensile-6.1.1.ebuild:
```diff
@@ -88,6 +88,8 @@ src_prepare() {
 }
 
 src_configure() {
+	rocm_use_hipcc
```
```suggestion
	use client && rocm_use_hipcc
```

negril commented:
@AngryLoki @littlewu2508 what's the status on this? Do you feel like it's ready for final checking and then merge?

ROCm 6.2.0 is out. And we should start working on that.

AngryLoki commented:
Please merge, I don't think that it needs extra polishing. If I'll notice anything I'll add to 6.2.0 PR.

littlewu2508 commented:
> what's the status on this? Do you feel like it's ready for final checking and then merge?

I feel the current ROCm-6.1 status with this PR is excellent.

One more thing: the `rocm.eclass` is updated, so @AngryLoki please send a v2 patch to gentoo-dev. Please squash the two rocm.eclass commit into one (since the second is a fix to the previous), and create the v2 patch with extra `-v 2 --interdiff=<the commit hash of v1 rocm.eclass patch>`, the reply to the first version of the patch.

> ROCm 6.2.0 is out. And we should start working on that.

Back in ROCm-5 I do not have enough time to bump all keep all ebuilds up-to-date with upstream. I guess it's becoming less painful these days, but still have to deal with a lot of patches and upstream changes. I'm OK if some versions are skipped, and prioritize fixing and polishing existing newest versions.

littlewu2508 commented on sci-libs/hipCUB/hipCUB-6.1.1.ebuild:
> Closes: https://bugs.gentoo.org/862837

Are you sure this is the fix to https://bugs.gentoo.org/862837? I guess that bug is just outdated, and irrelevant of this PR.

negril replied:
You can drop that, I closed the bug.

AngryLoki replied:
What I wanted to say, that in setups where systemwide compiler (e. g. clang-19) is configured as follows:
```
CFLAGS="-fuse-ld=gold"
CXXFLAGS="-fuse-ld=gold"
LDFLAGS="-fuse-linker-plugin"
```
and _other compiler_ works with these flags, while hipcc-needed one (e. g. clang-18) does not (for any reason, the reason does not matter), it throws away incompatible flags. Which ultimately fixes the issue. But now as bug is closed (and belongs to older clang where gold plugin was not build), I agree that this is not too relevant.

negril replied:
The gold plugin is also masked and unmaintained.

littlewu2508 commented:
I think this PR is ready, is there any reason not merging it?
